pkg/ls/run.go

```diff
@@ -0,0 +1,236 @@
+// ls: /c: No such file or directory
+// ls is more complex than it appears at first.
+// TODO: do we really need BOTH osfi and lsfi?
+// This may be required on non-unix systems like Plan 9 but it
+// would be nice to make sure.
+type file struct {
+	path string
+	osfi os.FileInfo
+	lsfi FileInfo
+	err  error
+}
+
+func (c cmd) listName(stringer Stringer, d string, prefix bool) error {
+	var files []file
+
+	filepath.Walk(d, func(path string, osfi os.FileInfo, err error) error {
+		f := file{
+			path: path,
+			osfi: osfi,
+		}
+
+		// error handling that matches standard ls is ... a real joy
+		if osfi != nil && !errors.Is(err, os.ErrNotExist) {
+			f.lsfi = FromOSFileInfo(path, osfi)
+			if err != nil && path == d {
+				f.err = err
+			}
+		} else {
+			f.err = err
+		}
+
+		files = append(files, f)
+
+		if err != nil {
+			return filepath.SkipDir
+		}
+
+		if !c.recurse && path == d && c.directory {
+			return filepath.SkipDir
+		}
+
+		if path != d && f.lsfi.Mode.IsDir() && !c.recurse {
+			return filepath.SkipDir
+		}
+
+		return nil
+	})
+
+	if c.size {
+		sort.SliceStable(files, func(i, j int) bool {
+			return files[i].lsfi.Size > files[j].lsfi.Size
+		})
+	}
+
+	for _, f := range files {
+		if f.err != nil {
+			c.printFile(stringer, f)
+			continue
+		}
+		if c.recurse {
+			// Mimic find command
+			f.lsfi.Name = f.path
+		} else if f.path == d {
+			if c.directory {
+				fmt.Fprintln(c.w, stringer.FileString(f.lsfi))
+				continue
+			}
+
+			// Starting directory is a dot when non-recursive
+			if f.osfi.IsDir() {
+				f.lsfi.Name = "."
+				if prefix {
+					if c.quoted {
+						fmt.Fprintf(c.w, "%q:\n", d)
+					} else {
+						fmt.Fprintf(c.w, "%v:\n", d)
+					}
+				}
+			}
+		}
+
+		c.printFile(stringer, f)
+	}
+
+	return nil
+}
+
+func indicator(fi FileInfo) string {
+	if fi.Mode.IsRegular() && fi.Mode&0o111 != 0 {
+		return "*"
+	}
+	if fi.Mode&os.ModeDir != 0 {
+		return "/"
+	}
+	if fi.Mode&os.ModeSymlink != 0 {
+		return "@"
+	}
+	if fi.Mode&os.ModeSocket != 0 {
+		return "="
+	}
+	if fi.Mode&os.ModeNamedPipe != 0 {
+		return "|"
+	}
+	return ""
+}
+
+func (c cmd) list(names []string) error {
+	if len(names) == 0 {
+		names = []string{"."}
+	}
+	// Write output in tabular form.
+	tw := &tabwriter.Writer{}
+	tw.Init(c.w, 0, 0, 1, ' ', 0)
+	c.w = tw
+	defer tw.Flush()
+
+	var s Stringer = NameStringer{}
+	if c.quoted {
+		s = QuotedStringer{}
+	}
+	if c.long {
+		s = LongStringer{Human: c.human, Name: s}
+	}
+	// Is a name a directory? If so, list it in its own section.
+	prefix := len(names) > 1
+	for _, d := range names {
+		if err := c.listName(s, d, prefix); err != nil {
+			return fmt.Errorf("error while listing %q: %w", d, err)
+		}
+		tw.Flush()
+	}
+	return nil
+}
+
+// Run executes ls with provided args and writes output to o.
+// args are as from os.Args where args[0] is the command name.
+func Run(o io.Writer, args []string) error {
+	var c cmd
+	f := flag.NewFlagSet(args[0], flag.ExitOnError)
```

Comment:
I don't think we should put flags in a package.

If someone wants to use the ls package, but wants l to mean something else, they are going to conflict.